### PR TITLE
[SDP-961] Change `POST /disbursements` to accept Verification Type

### PR DESCRIPTION
**Pull request by ziyliu**
### What
Add verification types to `/disbursement` endpoint

### Why
Finally leverage validation of verification type for the above endpoint

### Known limitations

[TODO or N/A]

#### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).
* [ ] This PR's title starts with the name of the package, area, or subject affected by the change.

#### Thoroughness

* [ ] This PR adds tests for the new functionality or fixes.
* [ ] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [ ] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [ ] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [ ] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [ ] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [ ] This is not a breaking change.
* [ ] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: internal/serve/validators/disbursement_validator.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/stellar-disbursement-platform-backend/internal/data"
+)
+
+const (
+	VERIFICATION_FIELD_PIN_MIN_LENGTH = 4
+	VERIFICATION_FIELD_PIN_MAX_LENGTH = 8
+
+	VERIFICATION_FIELD_MAX_ID_LENGTH = 50
```
Comment (ziyliu) [inline internal/serve/validators/disbursement_validator.go:15]: I checked with @torisamples on these values

**Review by marwen-abid — CHANGES_REQUESTED** [review]

File: internal/data/disbursements.go [hunk 2]
```diff
@@ -67,6 +67,19 @@ type DisbursementModel struct {
 	dbConnectionPool db.DBConnectionPool
 }
 
+type PostDisbursementRequest struct {
```
[the remark this answers is not in the document]
Comment (ziyliu) [reply internal/data/disbursements.go:70]: got it, good to know!

File: internal/data/disbursements.go [hunk 3]
```diff
@@ -67,6 +67,19 @@ type DisbursementModel struct {
...
+	Name              string            `json:"name"`
+	CountryCode       string            `json:"country_code"`
+	WalletID          string            `json:"wallet_id"`
+	AssetID           string            `json:"asset_id"`
+	VerificationValue string            `json:"verification_value"`
```
[the remark this answers is not in the document]
Comment (ziyliu) [reply internal/data/disbursements.go]: I've changed
- `POST` `/disbursements` to accept and store the `VerificationType` 
- seems like `POST` `disbursements/{id}/instructions` already calls `ValidateInstruction`, so I just needed to add the additional logic for the the 2 other types to the method!

File: internal/serve/httphandler/disbursement_handler_test.go [hunk 4]
```diff
@@ -97,7 +97,8 @@ func Test_DisbursementHandler_PostDisbursement(t *testing.T) {
 		{
 		  "wallet_id": "aab4a4a9-2493-4f37-9741-01d5bd31d68b",
 		  "asset_id": "61dbfa89-943a-413c-b862-a2177384d321",
-		  "country_code": "UKR"
+		  "country_code": "UKR",
+		  "verification_type": "date_of_birth"
```
Comment (marwen-abid) [inline internal/serve/httphandler/disbursement_handler_test.go]: shouldn't this be `verification_field` ?

File: internal/serve/httphandler/disbursement_handler_test.go [hunk 5]
```diff
@@ -142,20 +145,36 @@ func Test_DisbursementHandler_PostDisbursement(t *testing.T) {
...
 		requestBody, err := json.Marshal(PostDisbursementRequest{
 			Name:        "disbursement 1",
 			CountryCode: country.Code,
 			AssetID:     asset.ID,
-			WalletID:    "aab4a4a9-2493-4f37-9741-01d5bd31d68b",
+			WalletID:    enabledWallet.ID,
+		})
+		require.NoError(t, err)
+
+		want := `{"error":"Verification field invalid", "extras": {"verification_type": "invalid parameter. valid values are: DATE_OF_BIRTH, PIN, NATIONAL_ID_NUMBER"}}`
```
Comment (marwen-abid) [inline internal/serve/httphandler/disbursement_handler_test.go:166]: We should either go with `verification_type` or `verification field`. there is a mix of both in this error message.

**Review by marwen-abid — CHANGES_REQUESTED** [review]

File: internal/serve/httphandler/disbursement_handler_test.go (hunk 4 above)
Comment (ziyliu) [reply internal/serve/httphandler/disbursement_handler_test.go]: fixed this!

File: internal/serve/httphandler/disbursement_handler.go [hunk 6]
```diff
@@ -52,16 +54,21 @@ func (d DisbursementHandler) PostDisbursement(w http.ResponseWriter, r *http.Req
 		return
 	}
 
-	// validate request
-	v := validators.NewValidator()
+	iv := validators.NewDisbursementInstructionsValidator(disbursementRequest.VerificationField)
```
[the remark this answers is not in the document]
Comment (ziyliu) [reply internal/serve/httphandler/disbursement_handler.go:57]: @marwen-abid my thinking is that wouldn't we want to make sure that `VerificationField` is one of the three existing types? i.e. we decide to add a new verification type but not use it for this endpoint. The `ValidateAndGetVerificationType` just provides an additional error messaging for when the type is invalid.

[the remark this answers is not in the document]
Comment (marwen-abid) [reply internal/serve/httphandler/disbursement_handler.go:57]: I don't think this shortcut is worth it. 

the new method you created `ValidateAndGetVerificationType` is only used here, so I think it makes more sense to create a new validator called `DisbursementRequestValidator` that takes the whole disbursementRequest and validates it. You can move ValidateAndGetVerificationType to that new validator.

**Review by marwen-abid — APPROVED** [review]
Thank you! This looks great 🚀

**Pull request merged**